From ebcb460169c77b937aba7353538f320dc75ffd9e Mon Sep 17 00:00:00 2001
From: Dhanavandhana Kannan <quic_dhanavan@quicinc.com>
Date: Mon, 25 Mar 2024 11:55:41 +0530
Subject: [PATCH] hostapd: Add support to get the link mac address in BTM
 request frame

Currently for BSS Transition Management request frame, station's
MLD address will be sent via hostapd_cli command and that
address will be used to send the frame to the station.

Suppose if station's link address is given, it resuts in the
failure of the command as AP was not able to find the
station with the station's link address.

The fix is to provide support to get station link address
which is mentioned in hostapd_cli command for BSS Transition
Management request frame.

Signed-off-by: Dhanavandhana Kannan <quic_dhanavan@quicinc.com>
---
 src/ap/ctrl_iface_ap.c | 10 +++++++---
 src/ap/sta_info.c      | 12 ++++++++++++
 src/ap/sta_info.h      |  1 +
 3 files changed, 20 insertions(+), 3 deletions(-)

--- a/src/ap/ctrl_iface_ap.c
+++ b/src/ap/ctrl_iface_ap.c
@@ -1616,7 +1616,9 @@ int hostapd_ctrl_iface_bss_tm_req(struct
 	}
 
 	sta = ap_get_sta(hapd, addr);
-	if (sta == NULL) {
+	if (!sta && hapd->mld)
+		sta = ap_get_link_sta(hapd, addr);
+	if (!sta) {
 		wpa_printf(MSG_DEBUG, "Station " MACSTR
 			   " not found for BSS TM Request message",
 			   MAC2STR(addr));
--- a/src/ap/sta_info.c
+++ b/src/ap/sta_info.c
@@ -102,6 +102,19 @@ struct sta_info * ap_get_sta(struct host
 	return s;
 }
 
+struct sta_info *ap_get_link_sta(struct hostapd_data *hapd, const u8 *link_addr)
+{
+	struct sta_info *link_sta;
+
+	for (link_sta = hapd->sta_list; link_sta; link_sta = link_sta->next) {
+		if (link_sta->mld_info.mld_sta &&
+		    (os_memcmp(link_sta->mld_info.links[hapd->mld_link_id].peer_addr,
+		     link_addr, ETH_ALEN) == 0)) {
+			return link_sta;
+		}
+	}
+	return NULL;
+}
 
 #ifdef CONFIG_P2P
 struct sta_info * ap_get_sta_p2p(struct hostapd_data *hapd, const u8 *addr)
--- a/src/ap/sta_info.h
+++ b/src/ap/sta_info.h
@@ -370,6 +370,7 @@ int ap_for_each_sta(struct hostapd_data
 struct hostapd_ft_over_ds_ml_sta_entry *ap_get_ft_ds_ml_sta(struct hostapd_data *hapd,
 							    const u8 *sta_mld);
 struct sta_info * ap_get_sta(struct hostapd_data *hapd, const u8 *sta);
+struct sta_info *ap_get_link_sta(struct hostapd_data *hapd, const u8 *link_addr);
 struct sta_info * ap_get_sta_p2p(struct hostapd_data *hapd, const u8 *addr);
 void ap_sta_hash_add(struct hostapd_data *hapd, struct sta_info *sta);
 void ap_free_sta(struct hostapd_data *hapd, struct sta_info *sta);
